File: src/channels/feishu.ts
import { readFile } from "node:fs/promises"
import { resolve } from "node:path"

import { OrchestratorError } from "../errors.js"
import { requestJsonObject } from "../http.js"
import { tryParseJsonObject } from "../json-utils.js"
import { extractTextMessageContent, readObject } from "./feishu-core.js"

export interface FeishuRequirement {
  userId: string
  chatId: string
  messageId: string
  text: string
}

export interface FeishuAppCredentials {
  appId: string
  appSecret: string
  enabled: boolean
}

export async function loadFeishuCredentialsFromNanobot(
  configPath = "~/.nanobot/config.json",
): Promise<FeishuAppCredentials> {
  const absolutePath = resolve(configPath.replace(/^~(?=\/)/, process.env.HOME ?? "~"))
  let raw = ""
  try {
    raw = await readFile(absolutePath, "utf-8")
  } catch (error) {
    throw new OrchestratorError(`Nanobot config file not found: ${absolutePath}. ${String(error)}`)
  }

  let payload: unknown
  payload = tryParseJsonObject(raw)
  if (!payload) {
    throw new OrchestratorError(`Invalid Nanobot config JSON in ${absolutePath}`)
  }

  const channels = readObject(payload, "channels")
  const feishu = readObject(channels, "feishu")
  const appId = String(feishu.appId ?? feishu.app_id ?? "").trim()
  const appSecret = String(feishu.appSecret ?? feishu.app_secret ?? "").trim()
  const enabled = feishu.enabled !== false

  if (!appId || !appSecret) {
    throw new OrchestratorError("Nanobot feishu config missing appId/appSecret")
  }
  return { appId, appSecret, enabled }
}

export function parseRequirementEvent(payload: unknown): FeishuRequirement {
  const data = readObject(payload)
  const event = readObject(data, "event")
  const message = readObject(event, "message")
  const sender = readObject(event, "sender")

  const text = extractTextMessageContent(message.content, { fallbackToRawOnInvalidJson: false })
  if (!text) {
    throw new OrchestratorError("Feishu event does not contain text requirement")
  }

  const senderId =
    readObject(sender, "sender_id").open_id ?? sender.open_id ?? null
  const chatId = message.chat_id
  const messageId = message.message_id

  if (typeof senderId !== "string" || typeof chatId !== "string" || typeof messageId !== "string") {
    throw new OrchestratorError("Feishu event is missing sender/chat/message identifiers")
  }

  return {
    userId: senderId,
    chatId,
    messageId,
    text,
  }
}

export class FeishuMessenger {
  constructor(
    private readonly appId: string,
    private readonly appSecret: string,
    private readonly baseUrl = "https://open.feishu.cn/open-apis",
  ) {}

  async sendText(chatId: string, text: string): Promise<void> {
    const token = await this.tenantAccessToken()
    const response = await this.request("POST", "/im/v1/messages?receive_id_type=chat_id", {
      receive_id: chatId,
      msg_type: "text",
      content: JSON.stringify({ text }),
    }, token)

    if (response.code !== 0) {
      throw new OrchestratorError(`Failed to send Feishu message: ${JSON.stringify(response)}`)
    }
  }

  private async tenantAccessToken(): Promise<string> {
    const response = await this.request("POST", "/auth/v3/tenant_access_token/internal", {
      app_id: this.appId,
      app_secret: this.appSecret,
    })

    if (response.code !== 0 || typeof response.tenant_access_token !== "string") {
      throw new OrchestratorError(`Failed to fetch tenant token: ${JSON.stringify(response)}`)
    }
    return response.tenant_access_token
  }

  private async request(
    method: "POST" | "GET",
    path: string,
    payload?: Record<string, unknown>,
    token?: string,
  ): Promise<Record<string, unknown>> {
    const headers: Record<string, string> = {
      "content-type": "application/json; charset=utf-8",
    }
    if (token) {
      headers.authorization = `Bearer ${token}`
    }

  return requestJsonObject(`${this.baseUrl}${path}`, {
      method,
      headers,
      payload,
      timeoutMs: 10_000,
    })
  }
}

File: src/lucy_orchestrator/channels/feishu.py
from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from ..exceptions import OrchestratorError


@dataclass
class FeishuRequirement:
    user_id: str
    chat_id: str
    message_id: str
    text: str


@dataclass
class FeishuAppCredentials:
    app_id: str
    app_secret: str
    enabled: bool = True


def load_feishu_credentials_from_nanobot(
    config_path: str | Path = "~/.nanobot/config.json",
) -> FeishuAppCredentials:
    path = Path(config_path).expanduser().resolve()
    if not path.exists():
        raise OrchestratorError(f"Nanobot config file not found: {path}")

    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise OrchestratorError(f"Invalid Nanobot config JSON: {exc}") from exc

    channels = payload.get("channels", {}) if isinstance(payload, dict) else {}
    feishu = channels.get("feishu", {}) if isinstance(channels, dict) else {}
    if not isinstance(feishu, dict):
        raise OrchestratorError("Nanobot config does not contain channels.feishu")

    app_id = str(feishu.get("appId") or feishu.get("app_id") or "").strip()
    app_secret = str(feishu.get("appSecret") or feishu.get("app_secret") or "").strip()
    enabled = bool(feishu.get("enabled", True))

    if not app_id or not app_secret:
        raise OrchestratorError("Nanobot feishu config missing appId/appSecret")

    return FeishuAppCredentials(app_id=app_id, app_secret=app_secret, enabled=enabled)


def parse_requirement_event(payload: dict[str, Any]) -> FeishuRequirement:
    event = payload.get("event", {})
    message = event.get("message", {})
    sender = event.get("sender", {})

    content_raw = message.get("content", "{}")
    content: dict[str, Any]
    if isinstance(content_raw, str):
        content = json.loads(content_raw)
    else:
        content = dict(content_raw)

    text = str(content.get("text", "")).strip()
    if not text:
        raise OrchestratorError("Feishu event does not contain text requirement")

    sender_id = (
        sender.get("sender_id", {}).get("open_id") or sender.get("open_id") or ""
    )
    chat_id = message.get("chat_id", "")
    message_id = message.get("message_id", "")

    if not sender_id or not chat_id or not message_id:
        raise OrchestratorError(
            "Feishu event is missing sender/chat/message identifiers"
        )

    return FeishuRequirement(
        user_id=sender_id,
        chat_id=chat_id,
        message_id=message_id,
        text=text,
    )


class FeishuMessenger:
    def __init__(
        self,
        app_id: str,
        app_secret: str,
        base_url: str = "https://open.feishu.cn/open-apis",
    ) -> None:
        self.app_id = app_id
        self.app_secret = app_secret
        self.base_url = base_url.rstrip("/")

    def send_text(self, chat_id: str, text: str) -> None:
        token = self._tenant_access_token()
        payload = {
            "receive_id": chat_id,
            "msg_type": "text",
            "content": json.dumps({"text": text}, ensure_ascii=False),
        }
        response = self._request(
            method="POST",
            path="/im/v1/messages?receive_id_type=chat_id",
            payload=payload,
            token=token,
        )
        if response.get("code") != 0:
            raise OrchestratorError(f"Failed to send Feishu message: {response}")

    def _tenant_access_token(self) -> str:
        payload = {"app_id": self.app_id, "app_secret": self.app_secret}
        response = self._request(
            method="POST",
            path="/auth/v3/tenant_access_token/internal",
            payload=payload,
            token=None,
        )
        if response.get("code") != 0:
            raise OrchestratorError(f"Failed to fetch tenant token: {response}")
        token = response.get("tenant_access_token")
        if not token:
            raise OrchestratorError("Feishu token response missing tenant_access_token")
        return str(token)

    def _request(
        self,
        method: str,
        path: str,
        payload: dict[str, Any],
        token: str | None,
    ) -> dict[str, Any]:
        data = json.dumps(payload).encode("utf-8")
        request = urllib.request.Request(
            url=f"{self.base_url}{path}",
            method=method,
            data=data,
            headers={"Content-Type": "application/json; charset=utf-8"},
        )
        if token:
            request.add_header("Authorization", f"Bearer {token}")

        try:
            with urllib.request.urlopen(request, timeout=15) as response:
                body = response.read().decode("utf-8")
        except urllib.error.URLError as exc:
            raise OrchestratorError(f"Feishu request failed: {exc}") from exc

        parsed = json.loads(body)
        if not isinstance(parsed, dict):
            raise OrchestratorError("Invalid Feishu response payload")
        return parsed
